internal/sync/syncer.go

```diff
@@ -0,0 +1,311 @@
+func (s *Syncer) runOnce(ctx context.Context, key string, worker int) error {
+	log := logging.FromContext(ctx).With("worker", fmt.Sprint(worker)).With("gitSyncKey", key)
+	log.Debugf("Working on key: %s.", key)
+	strs := strings.Split(key, "/")
+	if len(strs) != 2 {
+		return fmt.Errorf("invalid key %q", key)
+	}
+	namespace := strs[0]
+	gitSyncName := strs[1]
+	gitSync := &v1alpha1.GitSync{}
+	if err := s.client.Get(ctx, client.ObjectKey{Namespace: namespace, Name: gitSyncName}, gitSync); err != nil {
+		if apierrors.IsNotFound(err) {
+			s.StopWatching(key)
+			log.Info("No corresponding GitSync found, stopped watching.")
+			return nil
+		}
+		return fmt.Errorf("failed to query GitSync object of key %q, %w", key, err)
+	}
+	if !gitSync.GetDeletionTimestamp().IsZero() {
+		s.StopWatching(key)
+		log.Debug("GitSync object being deleted.")
+		return nil
+	}
+
+	globalConfig, err := controllerConfig.GetConfigManagerInstance().GetConfig()
+	if err != nil {
+		log.Errorw("error getting  the  global config", "err", err)
+	}
+	repo, err := git.CloneRepo(ctx, s.client, gitSync, namespace, globalConfig)
+	if err != nil {
+		return fmt.Errorf("failed to clone the repo of key %q, %w", key, err)
+	}
+	manifests, err := git.GetLatestManifests(ctx, repo, gitSync)
+	if err != nil {
+		return fmt.Errorf("failed to get the manifest of key %q, %w", key, err)
+	}
+	uns, err := toUnstructuredAndApplyAnnotation(manifests, gitSyncName)
+	if err != nil {
+		return fmt.Errorf("failed to parse the manifest of key %q, %w", key, err)
+	}
+	synced := s.sync(gitSync, uns)
+	if synced {
+		log.Info("GitSync object is successfully synced.")
+	}
+	// TODO: commit the status
+
+	return err
+}
+
+type resourceInfoProviderStub struct {
+}
+
+func (r *resourceInfoProviderStub) IsNamespaced(_ schema.GroupKind) (bool, error) {
+	return false, nil
+}
+
+func (s *Syncer) sync(gitSync *v1alpha1.GitSync, targetObjs []*unstructured.Unstructured) bool {
+	// TODO: if the states match, then skip the syncing.
+	reconciliationResult, err := s.compareState(gitSync, targetObjs)
+	if err != nil {
+		return false
+	}
+
+	logEntry := log.WithFields(log.Fields{"gitsync": gitSync})
+
+	opts := []gitopssync.SyncOpt{
+		gitopssync.WithLogr(logging.NewLogrusLogger(logEntry)),
+		gitopssync.WithOperationSettings(false, true, false, false),
+		gitopssync.WithManifestValidation(true),
+		gitopssync.WithPruneLast(true),
+		gitopssync.WithReplace(false),
+		gitopssync.WithServerSideApply(true),
+		gitopssync.WithServerSideApplyManager(shared.SSAManager),
+	}
+
+	cluster, err := s.stateCache.GetClusterCache()
+	if err != nil {
+		return false
+	}
+	openAPISchema := cluster.GetOpenAPISchema()
+
+	syncCtx, cleanup, err := gitopssync.NewSyncContext(
+		gitSync.Spec.TargetRevision,
+		reconciliationResult,
+		s.config,
+		s.rawConfig,
+		s.kubectl,
+		gitSync.Spec.Destination.Namespace,
+		openAPISchema,
+		opts...,
+	)
+	defer cleanup()
+	if err != nil {
+		return false
+	}
+
+	syncCtx.Sync()
```

Comment:
Do we know if this is doing a Patch or Update under the hood and if there's a ResourceVersion check in the case that the reconciliationResult is now outdated?

Reply:
That is configured by the `replace` sync option.

Reply:
So, "replace==false" I guess means it's doing a Patch, right? And that Patch probably doesn't have a ResourceVersion in it, I imagine? So, there's some small risk that between calculating the reconciliationResult and applying it, something changes (say user adds a new field but our reconciliationResult doesn't override it). 

Back when we were doing client-side apply, auto-healing wouldn't take care of this issue because our "lastAppliedConfiguration" would forever be wrong. But since you're using Server Side Apply, auto-healing won't have that problem. We should be able to come back here, take the delta and then see that we need to override the field. So, as far as I can tell we should be good there.